### PR TITLE
CHAOS-577 - eBPF filters for network disruption without libbpfgo

### PR DESCRIPTION
**Pull request by aymericDD**
## What does this PR do?

- [x] Adds new functionality
- [ ] Alters existing functionality
- [ ] Fixes a bug
- [x] Improves documentation or testing

Please briefly describe your changes as well as the motivation behind them:
- Improve the network disruption by adding two tc eBPF filters:
  - **Path**: If the url match the path prefix the disruption will be applied
  - **Method**: If the the method of the url match the method filter it will be applied.
  - All filters must match in order to disrupt the request.
 
image

- **Update the documentation** to explain how the new tc eBPF filters are implemented:
filter_ebpf
filter_ebpf-2

## Code Quality Checklist

- [x] The documentation is up to date.
- [x] My code is sufficiently commented and passes continuous integration checks.
- [x] I have signed my commit (see [Contributing Docs](../CONTRIBUTING.md)).

## Testing

- [ ] I leveraged continuous integration testing
    - [ ] by depending on existing `unit` tests or `end-to-end` tests.
    - [ ] by adding new `unit` tests or `end-to-end` tests.
- [x] I manually tested the following steps:
    - [x] locally:
      - Deploy the demo application: `kubectl apply -f examples/namespace.yaml && kubectl apply -f examples/demo.yaml`
      -  Apply the network disruption `kubectl apply -f examples/network_drop.yaml`:
``` yaml
apiVersion: chaos.datadoghq.com/v1beta1
kind: Disruption
metadata:
  name: network-drop
  namespace: chaos-demo
  annotations:
    chaos.datadoghq.com/environment: "lima"
spec:
  level: pod
  selector:
    app: demo-curl
  count: 1
  network:
    drop: 100 # percentage of outgoing packets to drop
    method: get << New filter
    path: / << New filter
```

### CODE REVIEW

File: docs/network_disruption/prio.md [hunk 1]
```diff
@@ -152,6 +152,43 @@ We first filter for all packets related to `gateway IP` and `node IP` and send t
...
+  level: pod
+  selector:
+    app: demo
+  count: 1
+  network:
+    hosts:
+      - 10.0.1.254/31
+    port: 80
+    protocol: tcp
+    method: get <-- eBPF filter
```
Comment (ptnapoleon) [inline docs/network_disruption/prio.md:172]: Are these both HTTP specific? We don't ever seem to mention that in the docs or names or code.

File: docs/network_disruption/prio.md [hunk 2]
```diff
@@ -152,6 +152,43 @@ We first filter for all packets related to `gateway IP` and `node IP` and send t
...
+    path: /test <-- eBPF filter
```
Comment (ptnapoleon) [inline docs/network_disruption/prio.md:173]: This is a fairly sticky (as in hard to change) decision as to where to put these options. I think having only one each per disruption might be too inflexible? It's completely plausible to imagine someone wanting to have a small list of these, no?

`network` has always been L4 disruptions. If we're adding L7, do we not want them to be a different kind entirely?

File: docs/network_disruption/prio.md [hunk 3]
```diff
@@ -152,6 +152,43 @@ We first filter for all packets related to `gateway IP` and `node IP` and send t
 
 Finally, we apply a filter to enqueue all packets to class `1:4` whenever the `destination IP` is encompassed by the `hosts` field (see [this documentation](../../docs/network_disruption/hosts-and-services.md) for more details). In this case, a filter is applied for `10.0.1.254/32` and another for `10.0.1.255/32`. If no hosts were specified, a single filter is applied for `0.0.0.0/0` and no traffic is ends up in class `2:1`.
 
+### Network Disruption implementation for pod level with eBPF filters
+
+Tc can use eBPF filters in order to intercept packets at the application layer and allow the access to the payload of packets.
```
Comment (ptnapoleon) [inline docs/network_disruption/prio.md:157]: Are there requirements for the host we're disrupting on for these ebpf filters to work? I don't see documentation about that (unless it's in the images?)

Comment (aymericDD) [reply docs/network_disruption/prio.md:157]: Good point 👍  The host needs to have a Linux kernel with eBPF options.

File: docs/network_disruption/prio.md (hunk 1 above)
Comment (aymericDD) [reply docs/network_disruption/prio.md:172]: Yes, **method** and **path** are related to HTTP. Something like that can do the job?

``` yaml
spec:
  level: pod
  selector:
    app: demo
  count: 1
  network:
    hosts:
      - 10.0.1.254/31
    port: 80 <--- L4 
    protocol: tcp
    http:
      method: get <-- eBPF filter
      path: /test <-- eBPF filter
 ```

File: docs/network_disruption/prio.md (hunk 2 above)
Comment (aymericDD) [reply docs/network_disruption/prio.md:173]: I agree with you, it could be nice to be able to define a list of **path** and **method** to apply the disruption, but I just do something simple for first PR to simplify the review. Currently, it is not possible at all to filter on `method` and `path`. This is the same point for the disk failure injection based on eBPF.

What do you mean by a `different kind entirely`? Do you have an example?

Comment (ptnapoleon) [reply docs/network_disruption/prio.md:207]: `Kind` as in an entirely new top level disruption type, alongside network, e.g.
```
spec:
  network:
    ...
  http:
    ...
```
    it's just an idea though

File: docs/network_disruption/prio.md (hunk 1 above)
Comment (ptnapoleon) [reply docs/network_disruption/prio.md:206]: Yes, I think that is much clearer

File: docs/network_disruption/prio.md (hunk 2 above)
Comment (ptnapoleon) [reply docs/network_disruption/prio.md:207]: Yes, given the implementation here, I do think something like https://github.com/DataDog/chaos-controller/pull/741#discussion_r1266427453 is preferable

Comment (aymericDD) [reply docs/network_disruption/prio.md]: I have updated the PR to defined filters like https://github.com/DataDog/chaos-controller/pull/741#discussion_r1266427453 👍

File: api/v1beta1/network_disruption.go [hunk 4]
```diff
@@ -24,7 +25,10 @@ const (
 	// When not specifying an index for the hashtable created when we use u32 filters, the default id for this hashtable is 0x800.
 	// However, the maximum id being 0xFFF, we can only have 2048 different ids, so 2048 tc filters with u32.
 	// https://github.com/torvalds/linux/blob/v5.19/net/sched/cls_u32.c#L689-L690
-	MaximumTCFilters = 2048
+	MaximumTCFilters           = 2048
+	MaxNetworkPathCharacters   = 100
+	DefaultNetworkMethodFilter = "ALL"
+	DefaultNetworkPathFilter   = "/"
```
Comment (ptnapoleon) [inline api/v1beta1/network_disruption.go:31]: Could we rename these `DefaultHTTPMethodFilter` and `DefaultHTTPPathFilter`?

File: api/v1beta1/network_disruption.go [hunk 5]
```diff
@@ -327,6 +376,11 @@ func (s *NetworkDisruptionSpec) Format() string {
 	return networkDescription
 }
 
+// HasHTTPFilters return true if a custom method or path is defined, else return false
+func (s *NetworkDisruptionSpec) HasHTTPFilters() bool {
+	return s.HTTP.Method != DefaultNetworkMethodFilter || s.HTTP.Path != DefaultNetworkPathFilter
```
Comment (ptnapoleon) [inline api/v1beta1/network_disruption.go:381]:
```suggestion
	return s.HTTP != nil && (s.HTTP.Method != DefaultNetworkMethodFilter || s.HTTP.Path != DefaultNetworkPathFilter)
```

gotta check the HTTP pointer

File: bin/injector/Dockerfile [hunk 6]
```diff
@@ -30,7 +30,6 @@ COPY --from=binaries /usr/lib/x86_64-linux-gn[u] /usr/lib/x86_64-linux-gnu/
 COPY --from=binaries /usr/bin/test /bin/sh
 
 ARG TARGETARCH
-ENV BPF_DISK_FAILURE_NAME "bpf-disk-failure-${TARGETARCH}"
```
Comment (ptnapoleon) [inline bin/injector/Dockerfile:33]: seems unrelated?

File: network/exectuor.go [hunk 7]
```diff
@@ -0,0 +1,89 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
```
Comment (ptnapoleon) [inline network/exectuor.go:1]: Typo in this file name, should be `executor.go`

**Review by ptnapoleon — COMMENTED** [review]
Some change requests, but otherwise lgtm

File: bin/injector/Dockerfile (hunk 6 above)
Comment (aymericDD) [reply bin/injector/Dockerfile:33]: Indeed, this environment variable is never used by the diskfailure.

File: ebpf/network-tc-filter/injection.bpf.c [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+                i++;
+            }
+
+            // Null-terminate the path
+            path[path_length] = '\0';
+            break;
+        }
+    }
+
+    printt("PATH: %s", path);
```
Comment (clairecng) [inline ebpf/network-tc-filter/injection.bpf.c:108]: Would be nice to see logs from the ebpf programs in datadog (to do in a later PR probably)

Comment (aymericDD) [reply ebpf/network-tc-filter/injection.bpf.c:108]: Good point 👍 For now I can create a documentation to explain how to debug a tc eBPF program.

**Review by ptnapoleon — APPROVED** [review]

File: network/tc.go [hunk 9]
```diff
@@ -64,57 +62,31 @@ type TrafficController interface {
...
-	// early exit if dry-run mode is enabled
-	if e.dryRun {
-		return 0, "", nil
-	}
-
-	err := cmd.Run()
-	if err != nil {
-		err = fmt.Errorf("encountered error (%w) using args (%s): %s", err, args, stderr.String())
+func (t *tc) ConfigBPFFilter(cmd executor, args ...string) error {
+	if _, _, err := cmd.Run(args); err != nil {
```
Comment (clairecng) [inline network/tc.go:78]: you can remove the if condition and return err here

**Review by clairecng — APPROVED** [review]
one small nit otherwise 🎉
